lib/spack/spack/extensions.py

```diff
@@ -24,39 +24,47 @@ def extension_name(path):
         path (str): path where the extension resides
 
     Returns:
-        The extension name or None if path doesn't match the format
-        for Spack's extension.
+        The extension name. An exception is raised if path doesn't match
+        the expected format for a Spack command extension.
     """
-    regexp_match = re.search(extension_regexp, os.path.basename(path))
+    regexp_match = re.search(_extension_regexp, os.path.basename(path))
     if not regexp_match:
-        msg = "[FOLDER NAMING]"
-        msg += " {0} doesn't match the format for Spack's extensions"
-        tty.warn(msg.format(path))
-        return None
+        raise ExtensionNamingError(path)
     return regexp_match.group(1)
 
 
 def load_command_extension(command, path):
     """Loads a command extension from the path passed as argument.
 
     Args:
-        command (str): name of the command
+        command (str): name of the command (contains ``-``, not ``_``).
         path (str): base path of the command extension
 
     Returns:
-        A valid module object if the command is found or None
+        A valid module if found and loadable; None if not found. Module
+    loading exceptions are passed through.
     """
     extension = extension_name(path)
-    if not extension:
-        return None
 
     # Compute the name of the module we search, exit early if already imported
     cmd_package = '{0}.{1}.cmd'.format(__name__, extension)
+    # TODO: For consistency, the below conversion should utilize
+    #       spack.cmd.python_name(), but currently this would create a
+    #       circular relationship between spack.cmd and
+    #       spack.extensions.
     python_name = command.replace('-', '_')
     module_name = '{0}.{1}'.format(cmd_package, python_name)
     if module_name in sys.modules:
         return sys.modules[module_name]
```

Comment:
@alalazo, @scheibelp While resolving a discrepancy between handling internal Spack commands and extension commands with respect to the presence (or expected lack thereof) of hyphens vs underscores, I noted that `load_command_extension()` is not currently able to deal with the presence of a hyphen in the extension name. Specifically, the presence of one ensures either that a command will not be found, or that it is unable to use the `import` keyword for absolute imports (`spack.extensions`).

There is a relatively straightforward way to deal with this, namely that the extracted extension name should be pythonized before being used for subsequent paths and namespaces. For instance: an extension named `build-env` (say) with an eponymous command should have the directory / file structure `spack-build-env/build_env/cmd/build_env.py`. This would enable python-keyword imports from the `spack.extensions.build_env` to function as expected. Please let me know if I should proceed with the change.

Reply:
@scheibelp I should certainly be able to reproduce this, but I may not get to this until just before or just after today's telcon.

Reply:
Note that having a hyphen in the extension name and a hyphen in the command name are distinct (but related). The changes in https://github.com/spack/spack/commit/576839a48c36e7a171dc5015274b97e9c9cf186c?#diff-6eefb714b69bf94913569ab1d1255b42L58-R62 illustrate that previously, if the extension was named `pathutils` and the command was named `foo-path`, then the source for the command would be expected to be found in `spack-pathutils/pathutils/cmd/foo-path.py` instead of the correct `spack-pathutils/pathutils/cmd/foo_path.py`. The command function would still be expected to be named `foo_path()` in either case, however.

576839a48c36e7a171dc5015274b97e9c9cf186c corrects this behavior specifically (and tests for it), but you would be correct in observing that the extension name is not fixed in the same way, and a hyphen in same would be an issue for, say, importing non-command modules from within the extension using absolute rather than relative notation (`import spack.extensions.<extension>.mymod` from within a command implementation rather than `import ..mymod`). It would make sense to fix this latter misbehavior at the same time and correctly distinguish from the extension name as a name and the extension name as a path or namespace entity.

Reply:
https://github.com/spack/spack/pull/13635/commits/8abe7492e4e5efe5a7747ced9f39245a3c2ffdf3 deals correctly with hyphenated extension names (and tests same).